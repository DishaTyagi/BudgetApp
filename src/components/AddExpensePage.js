import React from 'react';

const AddExpensePage = () => (
    <div>
        hi! i am add expense component
    </div>
);

export default AddExpensePage;