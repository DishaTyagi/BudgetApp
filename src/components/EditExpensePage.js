import React from 'react';

const EditExpensePage = () => (
    <div>
        hi! this is edit expense component
    </div>
)
export default EditExpensePage;