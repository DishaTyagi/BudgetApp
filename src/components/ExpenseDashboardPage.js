import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        hi! i am expense Dashboard Component
    </div>
);

export default ExpenseDashboardPage;