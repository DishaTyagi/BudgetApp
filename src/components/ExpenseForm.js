import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';   //css for dates.

const now = moment();
console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
    state = {
        description : '',
        note: '',
        amount: '',
        createdAt: moment(),
        calendarFocused: false,
        error: '',
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( () => (
            {
                description,                
            }
        ))
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState( () => ({
            note
        }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){      //string match method to match regex with a string.
            this.setState( { amount } );
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){      //if created is empty i.e the user simply deletes the date, don't change the state then.
            this.setState(() => (
                {
                    createdAt
                }
            ))    
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(() => (
            {
                calendarFocused: focused,
            }
        ))
    }
    onSubmit = (e) => {     
        e.preventDefault();     //prevents from full page refreshing.
        if(!this.state.description || !this.state.amount){
            this.setState(() => ({
                error: 'Please provide description and amount'
            }))
        }else{
            this.setState( () => ({
                error: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount , 10),      //converting text amount to float(base 10)
                createdAt: this.state.createdAt.valueOf(),       //created is a moment obj. Here we converted it to time stamp.
                note: this.state.note
            })
        }
    }
    render(){
        return (
            <div>
                {this.state.error && <h3>{this.state.error}</h3>}

                <form onSubmit={this.onSubmit} >
                    <label>Description</label>
                    <input type = "text" 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} 
                        autoFocus={true} 
                    />
                    <br />

                    <label>Amount</label>    
                    <input 
                        type="text" 
                        value = {this.state.amount}
                        onChange = {this.onAmountChange}    
                    />
                    <br />
                    <SingleDatePicker 
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = { () => false }        //lets us select the dates in the past.
                    />
                    <br />
                    <label>Note</label>
                    <textarea 
                        value={this.state.note} 
                        onChange = {this.onNoteChange}
                    >
                    </textarea>

                    <button>Add expense</button>
                </form>
            </div>
        )
    }
}

//amt is taken as text to enforce decimal numbers with 2 significant figures.