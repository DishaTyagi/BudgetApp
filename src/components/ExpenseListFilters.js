import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters'

class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };
    onFocusChange = (focused) => {
        this.setState({
            calendarFocused: focused
        })
    };
    render(){
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange = { (e) => {
                    this.props.dispatch(setTextFilter(e.target.value))           //dispatch method is already available inside props of react components.
                } } />
        
                <select value={this.props.filters.sortBy} onChange = { (e) => {
                    if(e.target.value === 'date'){
                        this.props.dispatch(sortByDate());
                    }else{
                        this.props.dispatch(sortByAmount());
                    }
                } } >
                    <option value="date" >Date</option>
                    <option value="amount" >Amount</option>
                </select>
                <DateRangePicker 
                    startDate = {this.props.filters.startDate}
                    endDate = {this.props.filters.endDate}
                    onDatesChange = {this.onDatesChange}
                    focusedInput = {this.state.calendarFocused}       
                    onFocusChange = {this.onFocusChange}
                    showClearDates = {true}
                    numberOfMonths = {1}
                    isOutsideRange = { ()=> false}
                />

            </div>
        )
    }
}

// const ExpenseListFilters = (props) => (
//     <div>
//         <input type="text" value={props.filters.text} onChange = { (e) => {
//             props.dispatch(setTextFilter(e.target.value))           //dispatch method is already available inside props of react components.
//         } } />

//         <select value={props.filters.sortBy} onChange = { (e) => {
//             if(e.target.value === 'date'){
//                 props.dispatch(sortByDate());
//             }else{
//                 props.dispatch(sortByAmount());
//             }
//         } } >
//             <option value="date" >Date</option>
//             <option value="amount" >Amount</option>
//         </select>

//     </div>
// )

const mapStateToProps = (state) => {
    return {        //this obj used as props to the component mentioned in the sec arg. of connect
        filters: state.filters,
    }
}

export default connect (mapStateToProps)(ExpenseListFilters);