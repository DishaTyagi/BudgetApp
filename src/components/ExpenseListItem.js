import React from 'react';
import {removeExpense} from '../actions/expenses';
import { connect } from 'react-redux';

const ExpenseListItem =  ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <p>{description}</p>
        <p>{amount} - {createdAt} </p>
        <button onClick = { () => {
            dispatch(removeExpense(id));
        } } >Remove</button>
    </div>
)

export default connect()(ExpenseListItem);      //this makes dispatch available to the components mentioned here.