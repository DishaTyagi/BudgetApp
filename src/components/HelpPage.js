import React from 'react';

const HelpPage = () => (
    <div>
        This is the helpPage component.
    </div>
)
export default HelpPage;
