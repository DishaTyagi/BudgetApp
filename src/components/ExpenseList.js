import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList =  (props) => (    //unconnected ExpenseList(diff from the connected one that is exported below)
    <div>
        {
            props.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map( expense => {
                    return <ExpenseListItem key={expense.id} {...expense}/>
                } )
            )
        }
        
    </div>
);
const mapStateToProps = (state) => {       //this fn tells what info from the store the component needs to access. state is the state of store.
    return {        //this obj. can be accessed as props in the components mentioned in the second argument.
        expenses: selectExpenses(state.expenses, state.filters)
    }
};
export default connect( mapStateToProps )(ExpenseList);
// export default ConnectedExpenseList;