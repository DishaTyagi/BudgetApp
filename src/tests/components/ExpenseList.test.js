import React from 'react';
import moment from 'moment';
import { shallow } from 'enzyme';
import {ExpenseList} from '../../components/ExpenseList';

const expenses = [{
    id: '1',
    description: 'copy',
    note: '',
    amount: 300,
    createdAt: moment(0).subtract(4, 'days').valueOf()    //moment 0 if jan 1 1970. To go 4 days back, subtract it by 4 days. Since we want the createdAt val to be int, use valueOf()
} , {
    id: '2',
    description: 'pen',
    note: '',
    amount: 200,
    createdAt: moment(0).valueOf()    //or 0  
}, {
    id: '3',
    description: 'door',
    note: '',
    amount: 100,
    createdAt: moment(0).add(4, 'days').valueOf()       //4 days after Jan 1, 1970.
}, {
    id: '4',
    description: 'mouse',
    note: '',
    amount: 900,
    createdAt: moment(0).add(10, 'days').valueOf(),
}];

test('should render expenseList with expenses' , () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});