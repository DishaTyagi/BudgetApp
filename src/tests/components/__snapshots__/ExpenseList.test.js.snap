// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenseList with empty message 1`] = `<div />`;

exports[`should render expenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={300}
    createdAt={-345600000}
    description="copy"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={200}
    createdAt={0}
    description="pen"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={100}
    createdAt={345600000}
    description="door"
    id="3"
    key="3"
    note=""
  />
  <ExpenseListItem
    amount={900}
    createdAt={864000000}
    description="mouse"
    id="4"
    key="4"
    note=""
  />
</div>
`;
